src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/TextBasedPhysicalPlanPrinterImpl2.java

```diff
@@ -0,0 +1,354 @@
+		out.append( System.lineSeparator() );
+		if ( lop instanceof LogicalOpUnion ) {
+			printLogicalOpUnionForPhysicalOp( (LogicalOpUnion) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpMultiwayUnion ) {
+			printLogicalOpMultiwayUnionForPhysicalOp( (LogicalOpMultiwayUnion) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpTPAdd ) {
+			printLogicalOpTpAddForPhysicalOp( (LogicalOpTPAdd) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpTPOptAdd ) {
+			printLogicalOpTpOptAddForPhysicalOp( (LogicalOpTPOptAdd) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpBGPAdd ) {
+			printLogicalOpBGPAddForPhysicalOp( (LogicalOpBGPAdd) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpBGPOptAdd ) {
+			printLogicalOpBGPOptAddForPhysicalOp( (LogicalOpBGPOptAdd) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpGPAdd ) {
+			printLogicalOpGPAddForPhysicalOp( (LogicalOpGPAdd) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpGPOptAdd) {
+			printLogicalOpGPOptAddForPhysicalOp( (LogicalOpGPOptAdd) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpFilter ) {
+			printLogicalOpFilterForPhysicalOp( (LogicalOpFilter) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpGlobalToLocal ) {
+			printLogicalOpGlobalToLocalForPhysicalOp( (LogicalOpGlobalToLocal) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpJoin ) {
+			printLogicalOpJoinForPhysicalOp( (LogicalOpJoin) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpRightJoin ) {
+			printLogicalOpRightJoinForPhysicalOp( (LogicalOpRightJoin) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpLocalToGlobal ) {
+			printLogicalOpLocalToGlobalForPhysicalOp( (LogicalOpLocalToGlobal) lop, out, indentLevelStringForOpDetail );
+		}
+		else if ( lop instanceof LogicalOpRequest ) {
+			printLogicalOpRequestForPhysicalOp( (LogicalOpRequest) lop, out, indentLevelStringForOpDetail );
+		}
+		else {
+			throw new IllegalArgumentException( "Unexpected logical operator type: " + lop.getClass().getName() );
+		}
+	} 
+	
+	protected void printLogicalOpTpAddForPhysicalOp (final LogicalOpTPAdd lop, final PrintStream out, final String indentLevelStringForOpDetail) {
+		out.append( indentLevelStringForOpDetail + singleBase + "  - lop (" + lop.getID() + ") tpAdd" );
+		out.append( System.lineSeparator() );
+		printFederationMember( lop.getFederationMember(), indentLevelStringForOpDetail + singleBase, out );
+		printSPARQLGraphPattern( lop.getTP(), indentLevelStringForOpDetail + singleBase, out );
```

Comment:
These things don't belong into the function that is supposed to print the `- lop` line.

Reply:
I don't understand the meaning here.  I have the following output for example from this piece of code after printing `lop ... tpAdd`.

```
indexNestedLoop (4) 
        │  - lop (7) tpAdd
        │  - fm (1) TPFInterface server at http://fragments.dbpedia.org/2016-04/en
        │  - pattern (-803321067) (?P2  <http://www.global-example#researchTopic>  ?R)
        │
        └── req (2)
              - fm (1) TPFInterface server at http://fragments.dbpedia.org/2016-04/en
              - pattern (2003851450) (?P2  a  <http://www.global-example#Professor>)
```